cpp/src/arrow/util/rle_util.cc

```diff
@@ -0,0 +1,24 @@
+#include "arrow/util/rle_util.h"
+#include <algorithm>
+#include "arrow/builder.h"
+
+namespace arrow {
+namespace rle_util {
+
+int64_t FindPhysicalOffset(const int64_t* accumulated_run_lengths,
+                           int64_t physical_length, int64_t logical_offset) {
+  auto it = std::upper_bound(accumulated_run_lengths,
+                             accumulated_run_lengths + physical_length, logical_offset);
+  return std::distance(accumulated_run_lengths, it);
+}
+
+void AddArtificialOffsetInChildArray(ArrayData* array, int64_t offset) {
```

Comment:
You could make this return `Status` and do `RETURN_NOT_OK`. I know this is a testing function but I'd say it's probably better that the test wraps this function in `ARROW_CHECK_OK`